diff --git a/src/arch/x86/isa/decoder/x87.isa b/src/arch/x86/isa/decoder/x87.isa
index dc5840f61a..fd0bf26a5c 100644
--- a/src/arch/x86/isa/decoder/x87.isa
+++ b/src/arch/x86/isa/decoder/x87.isa
@@ -44,7 +44,8 @@ format WarnUnimpl {
             0x2: fcom();
             0x3: fcomp();
             0x4: Inst::FSUB1(Ed);
-            0x5: fsubr();
+            0x5: BasicOperate::magicInst({{/*Nothing*/}}, 
+                                               IsMagic);
             0x6: Inst::FDIV1(Ed);
             0x7: fdivr();
         }
diff --git a/src/cpu/SConscript b/src/cpu/SConscript
index 64c64903bd..ecb38b3633 100644
--- a/src/cpu/SConscript
+++ b/src/cpu/SConscript
@@ -69,6 +69,8 @@ DebugFlag('O3PipeView')
 DebugFlag('PCEvent')
 DebugFlag('Quiesce')
 DebugFlag('Mwait')
+DebugFlag('HW2')
+DebugFlag('Magic')
 
 CompoundFlag('ExecAll', [ 'ExecEnable', 'ExecCPSeq', 'ExecEffAddr',
     'ExecFaulting', 'ExecFetchSeq', 'ExecOpClass', 'ExecRegDelta',
diff --git a/src/cpu/StaticInstFlags.py b/src/cpu/StaticInstFlags.py
index 2e02833d1a..cde5c92650 100644
--- a/src/cpu/StaticInstFlags.py
+++ b/src/cpu/StaticInstFlags.py
@@ -100,4 +100,5 @@ class StaticInstFlags(Enum):
         "IsHtmStop",  # Stops (commits) a HTM transaction
         "IsHtmCancel",  # Explicitely aborts a HTM transaction
         "IsInvalid",  # An invalid instruction
+        "IsMagic" # magic instructions cool!
     ]
diff --git a/src/cpu/o3/commit.cc b/src/cpu/o3/commit.cc
index 1752fdc60a..db615fa2ca 100644
--- a/src/cpu/o3/commit.cc
+++ b/src/cpu/o3/commit.cc
@@ -66,6 +66,7 @@
 #include "params/BaseO3CPU.hh"
 #include "sim/faults.hh"
 #include "sim/full_system.hh"
+#include "debug/HW2.hh"
 
 namespace gem5
 {
@@ -1005,6 +1006,12 @@ Commit::commitInsts()
                 if (head_inst->isHtmStop())
                     htmStops[tid]++;
 
+
+                if (head_inst->isMagic()) {
+                    DPRINTF(HW2, "Magic instruction committed at cycle %lu\n", curTick());
+                    cpu->getProfilingStats(tid);
+                }
+
                 changedROBNumEntries[tid] = true;
 
                 // Set the doneSeqNum to the youngest committed instruction.
diff --git a/src/cpu/o3/commit.hh b/src/cpu/o3/commit.hh
index 4fff9fe892..8dfc231469 100644
--- a/src/cpu/o3/commit.hh
+++ b/src/cpu/o3/commit.hh
@@ -491,6 +491,9 @@ class Commit
         /** Number of cycles where the commit bandwidth limit is reached. */
         statistics::Scalar commitEligibleSamples;
     } stats;
+
+  public:
+    double getBranchMispredictsFromCommit() const { return stats.branchMispredicts.result(); }
 };
 
 } // namespace o3
diff --git a/src/cpu/o3/cpu.cc b/src/cpu/o3/cpu.cc
index e161d88daa..7aee7670be 100644
--- a/src/cpu/o3/cpu.cc
+++ b/src/cpu/o3/cpu.cc
@@ -60,6 +60,8 @@
 #include "sim/process.hh"
 #include "sim/stat_control.hh"
 #include "sim/system.hh"
+#include "debug/HW2.hh"
+#include "debug/Magic.hh"
 
 namespace gem5
 {
@@ -1477,5 +1479,27 @@ CPU::htmSendAbortSignal(ThreadID tid, uint64_t htm_uid,
     }
 }
 
+void
+CPU::getProfilingStats(ThreadID tid) const 
+{ 
+    double idle_cycles = cpuStats.idleCycles.value();
+    double total_cycles = baseStats.numCycles.value();
+    double total_insts = baseStats.numInsts.value();
+
+    double ipc = (total_cycles > 0) ? (total_insts / total_cycles) : 0.0;
+    double idle_rate = (total_cycles > 0) ? (idle_cycles / total_cycles) : 0.0;
+
+    double branchMispreds = commit.getBranchMispredictsFromCommit();
+    double branchPreds = fetch.getpredictedBranchesFromFetch();
+    double mpki = (total_insts > 0) ? ((double)branchMispreds / total_insts) * 1000.0 : 0.0;
+    double pred_acc = (branchPreds > 0) ? (1.0 - (branchMispreds / branchPreds)) : 0.0;
+
+    double blockCycles = iew.getblockCyclesFromIEW();
+    double stall_rate = (total_cycles > 0) ? (blockCycles / total_cycles) : 0.0;
+
+    DPRINTF(Magic, "\nMagic Profiler: \n> IPC=%f \n> Idle Rate=%f \n> Stall Rate=%f \n> MPKI=%f \n> Predictor Accuracy=%f\n\n", 
+        ipc, idle_rate, stall_rate, mpki, pred_acc);
+}
+
 } // namespace o3
 } // namespace gem5
diff --git a/src/cpu/o3/cpu.hh b/src/cpu/o3/cpu.hh
index cc238e532c..d497e54c37 100644
--- a/src/cpu/o3/cpu.hh
+++ b/src/cpu/o3/cpu.hh
@@ -590,6 +590,10 @@ class CPU : public BaseCPU
     // hardware transactional memory
     void htmSendAbortSignal(ThreadID tid, uint64_t htm_uid,
                             HtmFailureFaultCause cause) override;
+
+    // Profiling stats
+    void getProfilingStats(ThreadID) const;
+
 };
 
 } // namespace o3
diff --git a/src/cpu/o3/decode.cc b/src/cpu/o3/decode.cc
index ac728a2f2c..bb0cfdb001 100644
--- a/src/cpu/o3/decode.cc
+++ b/src/cpu/o3/decode.cc
@@ -50,6 +50,7 @@
 #include "debug/O3PipeView.hh"
 #include "params/BaseO3CPU.hh"
 #include "sim/full_system.hh"
+#include "debug/HW2.hh"
 
 // clang complains about std::set being overloaded with Packet::set if
 // we open up the entire namespace std
@@ -656,6 +657,11 @@ Decode::decodeInsts(ThreadID tid)
         DPRINTF(Decode, "[tid:%i] Processing instruction [sn:%lli] with "
                 "PC %s\n", tid, inst->seqNum, inst->pcState());
 
+        if(inst->isMagic()) {
+            DPRINTF(HW2, "Magic instruction decoded at cycle %lu [tid:%i, sn:%llu]\n",
+                    curTick(), tid, inst->seqNum);
+        }
+
         if (inst->isSquashed()) {
             DPRINTF(Decode, "[tid:%i] Instruction %i with PC %s is "
                     "squashed, skipping.\n",
diff --git a/src/cpu/o3/dyn_inst.hh b/src/cpu/o3/dyn_inst.hh
index 9cb3d3caa9..7b6970b928 100644
--- a/src/cpu/o3/dyn_inst.hh
+++ b/src/cpu/o3/dyn_inst.hh
@@ -558,6 +558,8 @@ class DynInst : public ExecContext, public RefCounted
     bool isCondCtrl()     const { return staticInst->isCondCtrl(); }
     bool isUncondCtrl()   const { return staticInst->isUncondCtrl(); }
     bool isSerializing()  const { return staticInst->isSerializing(); }
+    bool isMagic()  const { return staticInst->isMagic(); }
+
     bool
     isSerializeBefore() const
     {
diff --git a/src/cpu/o3/fetch.cc b/src/cpu/o3/fetch.cc
index b0bea842f5..b500638ed7 100644
--- a/src/cpu/o3/fetch.cc
+++ b/src/cpu/o3/fetch.cc
@@ -67,6 +67,7 @@
 #include "sim/eventq.hh"
 #include "sim/full_system.hh"
 #include "sim/system.hh"
+#include "debug/HW2.hh"
 
 namespace gem5
 {
@@ -1039,6 +1040,11 @@ Fetch::buildInst(ThreadID tid, StaticInstPtr staticInst,
 
     DPRINTF(Fetch, "[tid:%i] Instruction is: %s\n", tid,
             instruction->staticInst->disassemble(this_pc.instAddr()));
+    
+    if (instruction->isMagic()) {
+        DPRINTF(HW2, "Magic instruction fetched at cycle %lu [tid:%i, sn:%llu]\n",
+                curTick(), tid, seq);
+    }
 
 #if TRACING_ON
     if (trace) {
diff --git a/src/cpu/o3/fetch.hh b/src/cpu/o3/fetch.hh
index a6957a660a..914ffb67bf 100644
--- a/src/cpu/o3/fetch.hh
+++ b/src/cpu/o3/fetch.hh
@@ -579,6 +579,9 @@ class Fetch
         /** Rate of how often fetch was idle. */
         statistics::Formula idleRate;
     } fetchStats;
+
+    public:
+      double getpredictedBranchesFromFetch() const { return fetchStats.predictedBranches.value(); }
 };
 
 } // namespace o3
diff --git a/src/cpu/o3/iew.cc b/src/cpu/o3/iew.cc
index a8acb4c762..dce816b83f 100644
--- a/src/cpu/o3/iew.cc
+++ b/src/cpu/o3/iew.cc
@@ -57,6 +57,7 @@
 #include "debug/IEW.hh"
 #include "debug/O3PipeView.hh"
 #include "params/BaseO3CPU.hh"
+#include "debug/HW2.hh"
 
 namespace gem5
 {
@@ -870,6 +871,11 @@ IEW::dispatchInsts(ThreadID tid)
                 "IQ.\n",
                 tid, inst->pcState(), inst->seqNum, inst->threadNumber);
 
+        if(inst->isMagic()) {
+            DPRINTF(HW2, "Magic instruction dispatched at cycle %lu [tid:%i, sn:%llu]\n",
+                    curTick(), tid, inst->seqNum);
+        }
+
         // Be sure to mark these instructions as ready so that the
         // commit stage can go ahead and execute them, and mark
         // them as issued so the IQ doesn't reprocess them.
diff --git a/src/cpu/o3/iew.hh b/src/cpu/o3/iew.hh
index 4fe8227dcc..025f53fbbd 100644
--- a/src/cpu/o3/iew.hh
+++ b/src/cpu/o3/iew.hh
@@ -475,6 +475,9 @@ class IEW
         /** Average number of woken instructions per writeback. */
         statistics::Formula wbFanout;
     } iewStats;
+
+    public:
+      double getblockCyclesFromIEW() const { return iewStats.blockCycles.value(); }
 };
 
 } // namespace o3
diff --git a/src/cpu/o3/rename.cc b/src/cpu/o3/rename.cc
index 59a38aed45..55ea3e2a0d 100644
--- a/src/cpu/o3/rename.cc
+++ b/src/cpu/o3/rename.cc
@@ -51,6 +51,7 @@
 #include "debug/O3PipeView.hh"
 #include "debug/Rename.hh"
 #include "params/BaseO3CPU.hh"
+#include "debug/HW2.hh"
 
 namespace gem5
 {
@@ -657,6 +658,11 @@ Rename::renameInsts(ThreadID tid)
                 "Processing instruction [sn:%llu] with PC %s.\n",
                 tid, inst->seqNum, inst->pcState());
 
+        if (inst->isMagic()) {
+            DPRINTF(HW2, "Magic instruction renamed at cycle %lu [tid:%i, sn:%llu]\n",
+                    curTick(), tid, inst->seqNum);
+        }
+
         // Check here to make sure there are enough destination registers
         // to rename to.  Otherwise block.
         if (!renameMap[tid]->canRename(inst)) {
diff --git a/src/cpu/static_inst.hh b/src/cpu/static_inst.hh
index 78e47f4ed8..e8373c64d2 100644
--- a/src/cpu/static_inst.hh
+++ b/src/cpu/static_inst.hh
@@ -195,6 +195,7 @@ class StaticInst : public RefCounted, public StaticInstFlags
     bool isHtmStart() const { return flags[IsHtmStart]; }
     bool isHtmStop() const { return flags[IsHtmStop]; }
     bool isHtmCancel() const { return flags[IsHtmCancel]; }
+    bool isMagic() const { return flags[IsMagic]; }
 
     bool isInvalid() const { return flags[IsInvalid]; }
 
